### PR TITLE
feat: xcm

### PR DESCRIPTION
**Pull request by weichweich**
## fixes KILTProtocol/ticket#2084

include XCM on the KILT chain

## Checklist:

- [ ] I have verified that the code works
  - [ ] No panics! (checked arithmetic ops, no indexing `array[3]` use `get(3)`, ...)
- [ ] I have verified that the code is easy to understand
  - [ ] If not, I have left a well-balanced amount of inline comments
- [ ] I have [left the code in a better state](https://deviq.com/principles/boy-scout-rule)
- [ ] I have documented the changes (where applicable)

### CODE REVIEW

**Review by wischli — COMMENTED** [review]
I have a few questions and some minor remarks about the comments.

File: runtimes/peregrine/src/xcm_config.rs [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+impl pallet_xcm::Config for Runtime {
+	type Event = Event;
+	type SendXcmOrigin = EnsureXcmOrigin<Origin, LocalOriginToLocation>;
+	type XcmRouter = XcmRouter;
+	type ExecuteXcmOrigin = EnsureXcmOrigin<Origin, LocalOriginToLocation>;
+	type XcmExecuteFilter = Nothing;
+	// ^ Disable dispatchable execute on the XCM pallet.
+	// Needs to be `Everything` for local testing.
+	type XcmExecutor = XcmExecutor<XcmConfig>;
```
[the remark this answers is not in the document]
Comment (weichweich) [reply runtimes/peregrine/src/xcm_config.rs:108]: yeah it's from parity

File: runtimes/common/src/xcm_config.rs [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+	};
+}
+
+//Note: this might move to polkadot's xcm module.
+/// Deny executing the xcm message if it matches any of the Deny filter
+/// regardless of anything else. If it passes the Deny, and matches one of the
+/// Allow cases then it is let through.
+pub struct DenyThenTry<Deny, Allow>(PhantomData<Deny>, PhantomData<Allow>)
+where
+	Deny: ShouldExecute,
```
[the remark this answers is not in the document]
Comment (weichweich) [reply runtimes/common/src/xcm_config.rs:50]: Good point. Why do you constrain generics for a struct? Most of them are irrelevant for the struct no?

File: runtimes/spiritnet/src/xcm_config.rs [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+);
+
+impl pallet_xcm::Config for Runtime {
+	type Event = Event;
+	type SendXcmOrigin = EnsureXcmOrigin<Origin, LocalOriginToLocation>;
+	type XcmRouter = XcmRouter;
+	type ExecuteXcmOrigin = EnsureXcmOrigin<Origin, LocalOriginToLocation>;
+	type XcmExecuteFilter = Nothing;
+	// ^ Disable dispatchable execute on the XCM pallet.
+	// Needs to be `Everything` for local testing.
```
[the remark this answers is not in the document]
Comment (weichweich) [reply runtimes/spiritnet/src/xcm_config.rs]: ba5a4d2a0add7a769b7cad86fc69032e41fabd76

File: runtimes/spiritnet/src/xcm_config.rs [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+	type XcmExecutor = XcmExecutor<XcmConfig>;
+	type XcmTeleportFilter = Everything;
+	type XcmReserveTransferFilter = Nothing;
+	type Weigher = FixedWeightBounds<UnitWeightCost, Call, MaxInstructions>;
+	type LocationInverter = LocationInverter<Ancestry>;
+	type Origin = Origin;
+	type Call = Call;
+
+	const VERSION_DISCOVERY_QUEUE_SIZE: u32 = 100;
+	// ^ Override for AdvertisedXcmVersion default
```
[the remark this answers is not in the document]
Comment (weichweich) [reply runtimes/spiritnet/src/xcm_config.rs]: ba5a4d2a0add7a769b7cad86fc69032e41fabd76

File: runtimes/spiritnet/src/xcm_config.rs [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+/// This is the type we use to convert an (incoming) XCM origin into a local
+/// `Origin` instance, ready for dispatching a transaction with Xcm's
+/// `Transact`. There is an `OriginKind` which can biases the kind of local
+/// `Origin` it will become.
+pub type XcmOriginToTransactDispatchOrigin = (
+	// // Sovereign account converter; this attempts to derive an `AccountId` from the origin location
+	// // using `LocationToAccountId` and then turn that into the usual `Signed` origin. Useful for
+	// // foreign chains who want to have a local sovereign account on this chain which they control.
+	// SovereignSignedViaLocation<LocationToAccountId, Origin>,
+
```
[the remark this answers is not in the document]
Comment (weichweich) [reply runtimes/spiritnet/src/xcm_config.rs]: ba5a4d2a0add7a769b7cad86fc69032e41fabd76

File: runtimes/peregrine/src/xcm_config.rs [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+	type XcmTeleportFilter = Everything;
+	type XcmReserveTransferFilter = Nothing;
+	type Weigher = FixedWeightBounds<UnitWeightCost, Call, MaxInstructions>;
+	type LocationInverter = LocationInverter<Ancestry>;
+	type Origin = Origin;
+	type Call = Call;
+
+	const VERSION_DISCOVERY_QUEUE_SIZE: u32 = 100;
+	// ^ Override for AdvertisedXcmVersion default
+	type AdvertisedXcmVersion = pallet_xcm::CurrentXcmVersion;
```
[the remark this answers is not in the document]
Comment (weichweich) [reply runtimes/peregrine/src/xcm_config.rs]: ba5a4d2a0add7a769b7cad86fc69032e41fabd76

File: runtimes/common/src/xcm_config.rs [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+		Deny::should_execute(origin, message, max_weight, weight_credit)?;
+		Allow::should_execute(origin, message, max_weight, weight_credit)
+	}
+}
+
+/// Deny ReserveTransfer to the relay chain. Allow
+pub type Barrier = DenyThenTry<
+	DenyReserveTransferToRelayChain,
+	(
+		// // We don't allow anything from any sibling chain, therefore the following is not included here:
```
[the remark this answers is not in the document]
Comment (weichweich) [reply runtimes/common/src/xcm_config.rs]: ba5a4d2a0add7a769b7cad86fc69032e41fabd76

File: runtimes/common/src/xcm_config.rs [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+		max_weight: Weight,
+		weight_credit: &mut Weight,
+	) -> Result<(), ()> {
+		Deny::should_execute(origin, message, max_weight, weight_credit)?;
+		Allow::should_execute(origin, message, max_weight, weight_credit)
+	}
+}
+
+/// Deny ReserveTransfer to the relay chain. Allow
+pub type Barrier = DenyThenTry<
```
[the remark this answers is not in the document]
Comment (weichweich) [reply runtimes/common/src/xcm_config.rs]: ba5a4d2a0add7a769b7cad86fc69032e41fabd76

File: runtimes/common/src/xcm_config.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (weichweich) [reply runtimes/common/src/xcm_config.rs]: 533952c

File: runtimes/common/src/xcm_config.rs [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+		message: &mut Xcm<Call>,
+		max_weight: Weight,
+		weight_credit: &mut Weight,
+	) -> Result<(), ()> {
+		Deny::should_execute(origin, message, max_weight, weight_credit)?;
+		Allow::should_execute(origin, message, max_weight, weight_credit)
+	}
+}
+
+/// Deny ReserveTransfer to the relay chain. Allow
```
[the remark this answers is not in the document]
Comment (weichweich) [reply runtimes/common/src/xcm_config.rs]: ba5a4d2a0add7a769b7cad86fc69032e41fabd76

[the remark this answers is not in the document]
Comment (weichweich) [reply runtimes/common/src/xcm_config.rs]: ba5a4d2a0add7a769b7cad86fc69032e41fabd76

File: runtimes/common/src/xcm_config.rs [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+				} | TransferReserveAsset {
+					dest: MultiLocation {
+						parents: 1,
+						interior: Here
+					},
+					..
+				}
+			)
+		}) {
+			return Err(()); // Deny
```
[the remark this answers is not in the document]
Comment (weichweich) [reply runtimes/common/src/xcm_config.rs:114]: It's copied from parity. So not sure if I want to touch it. 🤔

File: runtimes/common/Cargo.toml [hunk 11]
```diff
@@ -58,6 +63,10 @@ std = [
   "sp-io/std",
   "sp-runtime/std",
   "sp-std/std",
+  "xcm/std",
+  "xcm-executor/std",
+  "xcm-builder/std",
+  "polkadot-parachain/std",
```
[the remark this answers is not in the document]
Comment (weichweich) [reply runtimes/common/Cargo.toml]: [735a3e0](https://github.com/KILTprotocol/kilt-node/pull/382/commits/735a3e0e017b3c3240c2d0db8bcd8e7add66ea81)

File: runtimes/peregrine/src/lib.rs [hunk 12]
```diff
@@ -206,17 +210,34 @@ impl cumulus_pallet_parachain_system::Config for Runtime {
...
 
+impl cumulus_pallet_xcmp_queue::Config for Runtime {
+	type Event = Event;
+	type XcmExecutor = XcmExecutor<XcmConfig>;
+	type ChannelInfo = ParachainSystem;
+	type VersionWrapper = ();
+	type ExecuteOverweightOrigin = EnsureRoot<AccountId>;
+	type ControllerOrigin = EnsureRoot<AccountId>;
+	type ControllerOriginConverter = XcmOriginToTransactDispatchOrigin;
+	type WeightInfo = ();
```
[the remark this answers is not in the document]
Comment (weichweich) [reply runtimes/peregrine/src/lib.rs]: [6046f3c](https://github.com/KILTprotocol/kilt-node/pull/382/commits/6046f3c04c02e205b0d5838f4700eaa62af0c1a9)

File: runtimes/spiritnet/src/lib.rs [hunk 13]
```diff
@@ -201,17 +205,34 @@ impl cumulus_pallet_parachain_system::Config for Runtime {
...
 
 impl parachain_info::Config for Runtime {}
 
 impl cumulus_pallet_aura_ext::Config for Runtime {}
 
+impl cumulus_pallet_xcmp_queue::Config for Runtime {
+	type Event = Event;
+	type XcmExecutor = XcmExecutor<XcmConfig>;
+	type ChannelInfo = ParachainSystem;
+	type VersionWrapper = ();
```
[the remark this answers is not in the document]
Comment (weichweich) [reply runtimes/spiritnet/src/lib.rs]: [6046f3c](https://github.com/KILTprotocol/kilt-node/pull/382/commits/6046f3c04c02e205b0d5838f4700eaa62af0c1a9)

File: runtimes/spiritnet/src/lib.rs [hunk 14]
```diff
@@ -201,17 +205,34 @@ impl cumulus_pallet_parachain_system::Config for Runtime {
...
+	type ExecuteOverweightOrigin = EnsureRoot<AccountId>;
+	type ControllerOrigin = EnsureRoot<AccountId>;
+	type ControllerOriginConverter = XcmOriginToTransactDispatchOrigin;
+	type WeightInfo = ();
```
[the remark this answers is not in the document]
Comment (weichweich) [reply runtimes/spiritnet/src/lib.rs]: [6046f3c](https://github.com/KILTprotocol/kilt-node/pull/382/commits/6046f3c04c02e205b0d5838f4700eaa62af0c1a9)

File: runtimes/peregrine/src/lib.rs [hunk 15]
```diff
@@ -206,17 +210,34 @@ impl cumulus_pallet_parachain_system::Config for Runtime {
...
 
 impl parachain_info::Config for Runtime {}
 
 impl cumulus_pallet_aura_ext::Config for Runtime {}
 
+impl cumulus_pallet_xcmp_queue::Config for Runtime {
+	type Event = Event;
+	type XcmExecutor = XcmExecutor<XcmConfig>;
+	type ChannelInfo = ParachainSystem;
+	type VersionWrapper = ();
```
[the remark this answers is not in the document]
Comment (weichweich) [reply runtimes/peregrine/src/lib.rs]: [6046f3c](https://github.com/KILTprotocol/kilt-node/pull/382/commits/6046f3c04c02e205b0d5838f4700eaa62af0c1a9)

File: runtimes/peregrine/src/xcm_config.rs [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+	RelayChainAsNative<RelayChainOrigin, Origin>,
+	// Native converter for sibling Parachains; will convert to a `SiblingPara` origin when
+	// recognized.
+	SiblingParachainAsNative<cumulus_pallet_xcm::Origin, Origin>,
+	// Native signed account converter; this just converts an `AccountId32` origin into a normal
+	// `Origin::Signed` origin of the same 32-byte value.
+	SignedAccountId32AsNative<RelayNetworkId, Origin>,
+	// Xcm origins can be represented natively under the Xcm pallet's Xcm origin.
+	XcmPassthrough<Origin>,
+);
```
[the remark this answers is not in the document]
Comment (weichweich) [reply runtimes/peregrine/src/xcm_config.rs]: ba5a4d2a0add7a769b7cad86fc69032e41fabd76

File: runtimes/common/Cargo.toml [hunk 17]
```diff
@@ -28,6 +28,10 @@ sp-core = {git = "https://github.com/paritytech/substrate", default-features = f
 sp-io = {git = "https://github.com/paritytech/substrate", default-features = false, branch = "polkadot-v0.9.24"}
 sp-runtime = {git = "https://github.com/paritytech/substrate", default-features = false, branch = "polkadot-v0.9.24"}
 sp-std = {git = "https://github.com/paritytech/substrate", default-features = false, branch = "polkadot-v0.9.24"}
+xcm = {git = "https://github.com/paritytech/polkadot", default-features = false, branch = "release-v0.9.24"}
+xcm-executor = {git = "https://github.com/paritytech/polkadot", default-features = false, branch = "release-v0.9.24"}
+xcm-builder = {git = "https://github.com/paritytech/polkadot", default-features = false, branch = "release-v0.9.24"}
+polkadot-parachain = {git = "https://github.com/paritytech/polkadot", default-features = false, branch = "release-v0.9.24"}
```
[the remark this answers is not in the document]
Comment (weichweich) [reply runtimes/common/Cargo.toml]: 735a3e0e017b3c3240c2d0db8bcd8e7add66ea81

File: runtimes/spiritnet/src/lib.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (weichweich) [reply runtimes/spiritnet/src/lib.rs]: [6046f3c](https://github.com/KILTprotocol/kilt-node/pull/382/commits/6046f3c04c02e205b0d5838f4700eaa62af0c1a9)

File: runtimes/peregrine/src/lib.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (weichweich) [reply runtimes/peregrine/src/lib.rs]: [6046f3c](https://github.com/KILTprotocol/kilt-node/pull/382/commits/6046f3c04c02e205b0d5838f4700eaa62af0c1a9)

File: runtimes/common/src/xcm_config.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (wischli) [reply runtimes/common/src/xcm_config.rs:115]: Ah okay, then let's keep as is.

[the remark this answers is not in the document]
Comment (wischli) [reply runtimes/common/src/xcm_config.rs:115]: Maybe we can reference in a comment where this code block was taken from? Could be helpful for future updates etc.

**Review by wischli — APPROVED** [review]
XCM::send(Multilocation(here), b"LGTM".into())

**Pull request merged**